src/backend/cdb/cdbmutate.c

```diff
@@ -2219,15 +2219,34 @@ collect_shareinput_producers(PlannerInfo *root, Plan *plan)
 	shareinput_walker((Node *) plan, &ctx);
 }
 
+enum { ROOT_SLICE = -1, SEGMENT_SLICE = 0, QUERY_DISPATCHER_SLICE = 1 };
+
 /* Some helper: implements a stack using List. */
 static void
-shareinput_pushmot(ApplyShareInputContext *ctxt, int motid)
+shareinput_pushmot(ApplyShareInputContext *ctxt, Motion *motion)
 {
-	ctxt->motStack = lcons_int(motid, ctxt->motStack);
+	int qds = SEGMENT_SLICE;
+
+	/* Top node of subplan should have a Flow node. */
+	Assert(motion->plan.lefttree);
+	Assert(motion->plan.lefttree->flow);
+
+	Flow *flow = motion->plan.lefttree->flow;
```

Comment:
This line should be moved upper (above asserts).

Reply:
> This line should be moved upper (above asserts).

it may cause segfault

Reply:
Then we should rewrite this.

Reply:
done

Reply:
OK.
I think we need to check the whole new code with pgindent after finish.

Reply:
> we need to check the whole new code with pgindent

done